rust/src/composepost.rs

```diff
@@ -28,6 +29,123 @@ use crate::treefile::Treefile;
 const RPMOSTREE_RPMDB_LOCATION: &str = "usr/share/rpm";
 const TRADITIONAL_RPMDB_LOCATION: &str = "var/lib/rpm";
 
+fn dir_renameat(
+    src: &openat::Dir,
+    srcname: &str,
+    destdir: &openat::Dir,
+    destname: &str,
+) -> Result<()> {
+    nix::fcntl::renameat(
+        Some(src.as_raw_fd()),
+        srcname,
+        Some(destdir.as_raw_fd()),
+        destname,
+    )?;
+    Ok(())
+}
+
+#[context("Moving {}", name)]
+fn dir_move_if_exists(src: &openat::Dir, dest: &openat::Dir, name: &str) -> Result<()> {
+    if src.exists(name)? {
+        dir_renameat(src, name, dest, name)?;
+    }
+    Ok(())
+}
+
+/// Initialize an ostree-oriented root filesystem.
+///
+/// This is hardcoded; in the future we may make more things configurable,
+/// but the goal is for all state to be in `/etc` and `/var`.
+#[context("Init rootfs")]
+pub(crate) fn compose_init_rootfs(
+    rootfs_dfd: &openat::Dir,
+    treefile: &mut Treefile,
+) -> CxxResult<()> {
+    use nix::fcntl::OFlag;
+    println!("Initializing rootfs");
+
+    // Unfortunately fchmod() doesn't operate on an O_PATH descriptor
+    let flag = OFlag::O_DIRECTORY | OFlag::O_CLOEXEC;
+    let nix_rootfs = nix::dir::Dir::openat(
+        rootfs_dfd.as_raw_fd(),
+        ".",
+        flag,
+        Mode::from_bits(0o755).unwrap(),
+    )?;
+
+    const TOPLEVEL_DIRS: &[&str] = &["dev", "proc", "run", "sys", "var", "sysroot"];
+    const TOPLEVEL_SYMLINKS: &[(&str, &str)] = &[
+        ("var/opt", "opt"),
+        ("var/srv", "srv"),
+        ("var/mnt", "mnt"),
+        ("var/roothome", "root"),
+        ("var/home", "home"),
+        ("run/media", "media"),
+        ("sysroot/ostree", "ostree"),
+    ];
+
+    nix::sys::stat::fchmod(nix_rootfs.as_raw_fd(), Mode::from_bits(0o755).unwrap())
+        .context("rootfs chown")?;
+
+    TOPLEVEL_DIRS
+        .par_iter()
+        .try_for_each(|&d| rootfs_dfd.ensure_dir(d, 0o755))?;
+    TOPLEVEL_SYMLINKS
+        .par_iter()
+        .try_for_each(|&(dest, src)| rootfs_dfd.symlink(src, dest))?;
+
+    if treefile.parsed.tmp_is_dir.unwrap_or_default() {
+        let tmp_mode = 0o1777;
+        rootfs_dfd.ensure_dir("tmp", tmp_mode)?;
+        nix::sys::stat::fchmodat(
+            Some(nix_rootfs.as_raw_fd()),
+            "tmp",
+            Mode::from_bits(tmp_mode).unwrap(),
+            nix::sys::stat::FchmodatFlags::NoFollowSymlink,
+        )?;
+    } else {
+        rootfs_dfd.symlink("tmp", "sysroot/tmp")?;
+    }
+
+    Ok(())
+}
+
+// Initialize a basic root filesystem in @target_root_dfd, then walk over the
+// root filesystem in @src_rootfs_fd and take the basic content and perform some
+// core conversions from RPM conventions to OSTree conventions. For example:
+//  * Checksum the kernel in /boot (and move kernel /usr/lib/modules)
+//  * Migrate content in /var to systemd-tmpfiles
```

Comment:
It think this specific bullet item was already wrong in the C code, there is no tmpfiles.d handling in this function.